subsys/power/reboot.c

```diff
@@ -33,3 +36,45 @@ void sys_reboot(int type)
 		k_cpu_idle();
 	}
 }
+
+#ifdef CONFIG_ARDUINO_LIKE_UPGRADE
+
+void sys_reboot_to_upgrade(int delay_ms)
+{
+	k_sleep(Z_TIMEOUT_MS(delay_ms));
+
+	(void)irq_lock();
+#ifdef CONFIG_SYS_CLOCK_EXISTS
+	sys_clock_disable();
+#endif
+
+#ifdef CONFIG_SOC_SERIES_SAMD21
```

Comment:
I like the goal of the PR, but I think this needs better abstraction / layering. I don't know this subsystem well enough to say exactly how it should go, though.

Reply:
I agree that the abstraction here is not ideal but I don't really see a good alternative other than creating a totally new subsystem for enabling booting into a arduino-compatible samd bootloader which seems like an overkill.

I didn't want to allow much code to run between the erasure of the first line of code section and the reset and so I made the two take place in the same routine. And as it essentially creates a slightly different-flavoured kind or reset, I thought it would be appropriate it lives next to the vanilla reset code.

Reply:
To be honest, I very much doubt this PR will be merged without a better abstraction, but I'll remove my -1 for the Python bits.